ion/cmp_test.go

```diff
@@ -0,0 +1,229 @@
+	return cmp.Equal(thisInt64, other)
+}
+
+func (thisUint64 ionUint64) eq(other ionEqual) bool {
+	return cmp.Equal(thisUint64, other)
+}
+
+func (thisBigInt ionBigInt) eq(other ionEqual) bool {
+	if val, ok := other.(ionBigInt); ok {
+		return thisBigInt.Int.Cmp(val.Int) == 0
+	}
+	return false
+}
+
+func (thisBool ionBool) eq(other ionEqual) bool {
+	return cmp.Equal(thisBool, other)
+}
+
+func (thisStr ionString) eq(other ionEqual) bool {
+	return cmp.Equal(thisStr, other)
+}
+
+func (thisTimestamp ionTimestamp) eq(other ionEqual) bool {
+	if val, ok := other.(ionTimestamp); ok {
+		return thisTimestamp.Time.Equal(val.Time)
+	}
+	return false
+}
+
+func (thisDecimal ionDecimal) eq(other ionEqual) bool {
+	return cmp.Equal(thisDecimal, other)
+}
+
+func cmpAnnotations(thisAnnotations, otherAnnotations []string) bool {
+	if len(thisAnnotations) != len(otherAnnotations) {
+		return false
+	}
+
+	for idx, this := range thisAnnotations {
+		other := otherAnnotations[idx]
+		thisText, otherText := this, other
+
+		if !cmp.Equal(thisText, otherText) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func cmpValueSlices(thisValues, otherValues []interface{}) bool {
+	if len(thisValues) == 0 && len(otherValues) == 0 {
+		return true
+	}
+
+	if len(thisValues) != len(otherValues) {
+		return false
+	}
+
+	res := false
+	for idx, this := range thisValues {
+		other := otherValues[idx]
+		thisType, otherType := reflect.TypeOf(this), reflect.TypeOf(other)
+
+		if thisType != otherType {
+			return false
+		}
+
+		switch this.(type) {
+		case string: // null.Sexp, null.List, null.Struct
+			res = strNullTypeCmp(this, other)
+		default:
+			thisItem := this.(ionItem)
+			otherItem := other.(ionItem)
+			res = thisItem.equal(otherItem)
+		}
+		if !res {
+			return false
+		}
+	}
+	return res
+}
+
+func cmpFloats(thisValue, otherValue interface{}) bool {
+	if !haveSameTypes(thisValue, otherValue) {
+		return false
+	}
+
+	switch val := thisValue.(type) {
+	case string: // null.float
+		return strNullTypeCmp(val, otherValue)
+	case float64:
+		thisFloat := ionFloat(val)
+		return thisFloat.eq(ionFloat(otherValue.(float64)))
+	default:
+		return false
+	}
+}
+
+func cmpInts(thisValue, otherValue interface{}) bool {
+	if !haveSameTypes(thisValue, otherValue) {
+		return false
+	}
+
+	switch val := thisValue.(type) {
+	case string: // null.int
+		return strNullTypeCmp(val, otherValue)
+	case int:
+		thisInt := ionInt(val)
+		return thisInt.eq(ionInt(otherValue.(int)))
+	case int64:
+		thisInt := ionInt64(val)
+		return thisInt.eq(ionInt64(otherValue.(int64)))
+	case uint64:
+		thisInt := ionUint64(val)
+		return thisInt.eq(ionUint64(otherValue.(uint64)))
+	case *big.Int:
+		thisInt := ionBigInt{val}
+		return thisInt.eq(ionBigInt{otherValue.(*big.Int)})
+	default:
+		return false
+	}
+}
+
+func cmpBools(thisValue, otherValue interface{}) bool {
+	if !haveSameTypes(thisValue, otherValue) {
+		return false
+	}
+
+	switch val := thisValue.(type) {
```

Comment:
I don't think I understand what this is doing. `thisValue` doesn't specify a type in the function signature, but we expect `thisValue.(type)` to sometimes be `string`, which represents `null.bool`. Could you walk me through it?

Reply:
if ionType of the value under test is deduced to BoolType([here](https://github.com/Armanbqt/ion-go/pull/9/files#diff-a44fc38e58a691d2b9876dfeefd76ae7R52)), we will get to this helper function.
But aside from "true/false" Ion bool value can also be `null.bool`, which is a string for Go compiler (but still bool for Ion).